sonar-db/src/main/java/org/sonar/db/charset/MssqlCharsetHandler.java

```diff
@@ -46,97 +42,81 @@
 
-  protected MssqlCharsetHandler(SqlExecutor selectExecutor) {
+  private final Metadata metadata;
+
+  MssqlCharsetHandler(SqlExecutor selectExecutor) {
     super(selectExecutor);
+    this.metadata = new Metadata();
+  }
+
+  @VisibleForTesting
+  MssqlCharsetHandler(SqlExecutor selectExecutor, Metadata metadata) {
+    super(selectExecutor);
+    this.metadata = metadata;
   }
 
   @Override
-  void handle(Connection connection, Set<DatabaseCharsetChecker.Flag> flags) throws SQLException {
-    logInit(flags);
+  void handle(Connection connection, DatabaseCharsetChecker.State state) throws SQLException {
+    if (state == DatabaseCharsetChecker.State.FRESH_INSTALL) {
+      expectCaseSensitiveDefaultCollation(connection);
+    } else if (state == DatabaseCharsetChecker.State.UPGRADE) {
+      expectCaseSensitiveDefaultCollation(connection);
+      repairColumns(connection);
+    }
+  }
+
+  private void expectCaseSensitiveDefaultCollation(Connection connection) throws SQLException {
+    LOGGER.info("Verify that database collation is case-sensitive and accent-sensitive");
+    String defaultCollation = metadata.getDefaultCollation(connection);
+
+    if (!isCollationCorrect(defaultCollation)) {
+      String fixedCollation = toCaseSensitive(defaultCollation);
+      throw MessageException.of(format(
+        "Database collation must be case-sensitive and accent-sensitive. It is %s but should be %s.", defaultCollation, fixedCollation));
+    }
+  }
+
+  private void repairColumns(Connection connection) throws SQLException {
+    String defaultCollation = metadata.getDefaultCollation(connection);
 
     // All VARCHAR columns are returned. No need to check database general collation.
     // Example of row:
     // issues | kee | Latin1_General_CS_AS or Latin1_General_100_CI_AS_KS_WS
-    Set<String> errors = new LinkedHashSet<>();
-    List<ColumnDef> columns = select(connection,
-      ColumnDef.SELECT_COLUMNS +
-        "FROM [INFORMATION_SCHEMA].[COLUMNS] " +
-        "WHERE collation_name is not null " +
-        "ORDER BY table_name,column_name",
-      ColumnDef.ColumnDefRowConverter.INSTANCE);
-    for (ColumnDef column : from(columns).filter(ColumnDef.IsInSonarQubeTablePredicate.INSTANCE)) {
-      if (!isCollationCorrect(column)) {
-        if (flags.contains(AUTO_REPAIR_COLLATION)) {
-          repairColumnCollation(connection, column);
-        } else {
-          errors.add(format("%s.%s", column.getTable(), column.getColumn()));
-        }
+    List<ColumnDef> columns = metadata.getColumnDefs(connection);
+    for (ColumnDef column : columns.stream().filter(ColumnDef::isInSonarQubeTable).collect(Collectors.toList())) {
+      String collation = column.getCollation();
+      if (!isCollationCorrect(collation)) {
+        repairColumnCollation(connection, column, toCaseSensitive(collation));
+      } else if (collation.equals("Latin1_General_CS_AS") && !collation.equals(defaultCollation)) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move the "Latin1_General_CS_AS" string literal on the left side of this string comparison. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1132)